internal/receiver/awscontainerinsightreceiver/cadvisor/cadvisor_linux.go

```diff
@@ -0,0 +1,224 @@
+			logger.Warn("Expected cAdvisor flag not found", zap.String("name", name))
+		}
+	}
+}
+
+func New(containerOrchestrator string, logger *zap.Logger) *Cadvisor {
+	nodeName := os.Getenv("HOST_NAME")
+	if nodeName == "" {
+		err := errors.New("Missing environment variable HOST_NAME. Please check your YAML config.")
+		logger.Warn("Error in Cadvisor initialization: ", zap.Error(err))
+	}
+
+	c := &Cadvisor{
+		logger:                logger,
+		nodeName:              nodeName,
+		version:               "0",
+		containerOrchestrator: containerOrchestrator,
+	}
+	overrideCadvisorFlagDefault(logger)
+
+	if err := c.initManager(); err != nil {
+		c.logger.Warn("Fail to initialize cadvisor", zap.Error(err))
+		return nil
+	}
+
+	c.machineInfo = host.NewMachineInfo(c.manager, time.Minute, c.logger)
+	return c
+}
+
+var metricsExtractors = []extractors.MetricExtractor{}
+
+func GetMetricsExtractors() []extractors.MetricExtractor {
+	return metricsExtractors
+}
+
+func (c *Cadvisor) addEbsVolumeInfo(tags map[string]string, ebsVolumeIdsUsedAsPV map[string]string) {
+	deviceName, ok := tags[common.DiskDev]
+	if !ok {
+		return
+	}
+
+	if c.ebsVolume != nil {
+		if volId := c.ebsVolume.GetEbsVolumeId(deviceName); volId != "" {
+			tags[common.HostEbsVolumeId] = volId
+		}
+	}
+
+	if tags[common.MetricType] == common.TypeContainerFS || tags[common.MetricType] == common.TypeNodeFS ||
+		tags[common.MetricType] == common.TypeNodeDiskIO || tags[common.MetricType] == common.TypeContainerDiskIO {
+		if volId := ebsVolumeIdsUsedAsPV[deviceName]; volId != "" {
+			tags[common.EbsVolumeId] = volId
+		}
+	}
+}
+
+func (c *Cadvisor) decorateMetrics(cadvisormetrics []*extractors.CAdvisorMetric) {
+	if c.ebsVolume == nil {
+		//delay the initialization. If instance id is not available, c.ebsVolume is set to nil
+		c.ebsVolume = host.NewEbsVolume(c.machineInfo.GetInstanceID(), time.Minute, c.logger)
+	}
+
+	ebsVolumeIdsUsedAsPV := host.ExtractEbsIdsUsedByKubernetes()
+
+	for _, m := range cadvisormetrics {
+		tags := m.GetTags()
+		c.addEbsVolumeInfo(tags, ebsVolumeIdsUsedAsPV)
+
+		//add version
+		tags[common.Version] = c.version
+
+		//add NodeName for node, pod and container
+		metricType := tags[common.MetricType]
+		if c.nodeName != "" && (common.IsNode(metricType) || common.IsInstance(metricType) ||
+			common.IsPod(metricType) || common.IsContainer(metricType)) {
+			tags[common.NodeNameKey] = c.nodeName
+		}
+
+		//add instance id and type
+		if instanceId := c.machineInfo.GetInstanceID(); instanceId != "" {
+			tags[common.InstanceId] = instanceId
+		}
+		if instanceType := c.machineInfo.GetInstanceType(); instanceType != "" {
+			tags[common.InstanceType] = instanceType
+		}
+	}
+}
+
+func (c *Cadvisor) GetMetrics() []pdata.Metrics {
+	c.logger.Debug("collect data from cadvisor...")
+	var result []pdata.Metrics
+	var containerinfos []*cinfo.ContainerInfo
+	var err error
+
+	if c.manager == nil && c.initManager() != nil {
+		panic("Cannot initiate manager")
+	}
```

Comment:
Q: do we need to call `c.initManager()` for every refresh? can we move it into `receiver.Start()`?